### PR TITLE
Make it possible for a screen to set gesturesEnabled

### PR DESCRIPTION
**Pull request by Ashoat**
Context for the problem I was facing #383. A related `gesturesEnabled` PR with some additional context in #292.

In StackNavigator, sometimes it's useful to prevent somebody from dismissing a route. My use case is a log-in screen that I display as a modal above the rest of my app.

Test plan: I hooked this into my own app, and verified that I was able to set gesturesEnabled via navigation options. flow/jest pass

### CODE REVIEW

**Comment by Ashoat** [comment]
Note that I removed the `gesturesEnabled` prop from `CardStack`. The thinking was that it would be set via navigationOptions (either as a static member of the screen, or the defaults specified in the StackNavigator config), and there wasn't any existing use case for setting it as a prop.

File: docs/api/navigators/StackNavigator.md [hunk 1]
```diff
@@ -125,6 +125,7 @@ All `navigationOptions` for the `StackNavigator`:
   - `style` - Style object for the navigation bar
   - `titleStyle` - Style object for the title component
   - `tintColor` - Tint color for the header
+- `gesturesEnabled` - Whether you can use gestures to dismiss this screen. Defaults to true on iOS, false on Android
```
Comment (satya164) [inline docs/api/navigators/StackNavigator.md]: nit: indent

File: src/TypeDefinition.js [hunk 2]
```diff
@@ -185,6 +185,11 @@ export type NavigationScreenOptions = {
    * Options passed to the drawer for this screen.
    */
   drawer?: NavigationScreenOption<DrawerConfig>;
+  /**
+   * Whether you can use gestures to dismiss this screen.
+   * Defaults to true on iOS, false on Android.
+   */
+  gesturesEnabled?: bool,
```
Comment (satya164) [inline src/TypeDefinition.js]: can you put it inside another object `cardStack`? we'll likely add more options in future and it'll be useful to access navigation state and params.

**Comment by Ashoat** [comment]
Okay, I encapsulated the new config param with a `cardStack` config object.

(The indentation nit you pointed out on the documentation is on purpose; check the preview)

File: src/views/CardStack.js [hunk 3]
```diff
@@ -343,7 +331,16 @@ class CardStack extends Component<DefaultProps, Props, void> {
...
     let panHandlers = null;
 
-    if (this.props.gesturesEnabled) {
+    const cardStackConfig = this.props.router.getScreenConfig(
+      props.navigation,
+      'cardStack'
+    ) || {};
+    const gesturesEnabledConfig = cardStackConfig.gesturesEnabled;
+    const gesturesEnabled =
+      (gesturesEnabledConfig === true || gesturesEnabledConfig === false)
```
Comment (satya164) [inline src/views/CardStack.js]: `typeof gesturesEnabledConfig === 'boolean'`

Comment (Ashoat) [inline src/views/CardStack.js]: Okay, done. Also updated the styling of the ternary expression to match what was used elsewhere in the file

**Comment by satya164** [comment]
cc @ericvicenti

File: src/views/CardStack.js [hunk 4]
```diff
@@ -343,7 +331,15 @@ class CardStack extends Component<DefaultProps, Props, void> {
...
 
-    if (this.props.gesturesEnabled) {
+    const cardStackConfig = this.props.router.getScreenConfig(
+      props.navigation,
+      'cardStack'
+    ) || {};
+    const gesturesEnabledConfig = cardStackConfig.gesturesEnabled;
+    const gesturesEnabled = typeof gesturesEnabledConfig === 'boolean' ?
+      gesturesEnabledConfig :
+      Platform.OS === 'ios';
```
Comment (ericvicenti) [inline src/views/CardStack.js]: This logic looks good to me, but can you write a comment describing the platform defined default?

**Comment by Ashoat** [comment]
Okay, added a comment.

One thought I had - would this config parameter be better named `dismissGesturesEnabled`? I think it's a bit more descriptive of what it actually controls, although if there are future plans to add non-dismiss gestures to `Card`/`CardStack` then it wouldn't really be helpful.

**Comment by satya164** [comment]
I think `gesturesEnabled` is fine.

**Comment by RichardLindhout** [comment]
If you want to do this on your login screen. You can do something like this.
``` 
import {
  StackNavigator as stackNavigator,
  

} from 'react-navigation'


import Login from './routes/Login'
import Home from './routes/Home'

import EditTimer from './routes/EditTimer'

const MainNavigator = stackNavigator(
  {
    Home: { screen: Home },
    EditTimer: { screen: EditTimer },
  },
  {
    cardStyle: {

      opacity: 1,
    },
    navigationOptions: {


      header: {
        tintColor: 'rgba(255,255,255,0.8)',
        titleStyle: {
          color: 'rgba(255,255,255,0.8)',
          fontWeight: '400',
        },
        style: {
          backgroundColor: 'rgba(69,37,200,0.8)',
          shadowColor: 'rgba(69,37,200,0.8)',
          shadowOpacity: 1,
          shadowRadius: 3,
          shadowOffset: {
            height: 3,
            width: 0,
          },
        },
      },
    },
    headerMode: 'float',
  }
)

const Router = stackNavigator(
  {
    Main: { screen: MainNavigator },
    Login: { screen: Login },
  },
  {
    headerMode: 'none',
    mode: 'modal',
    cardStyle: {
      opacity: 1,
    },
    navigationOptions: {
      cardStack: {
        gesturesEnabled: true,
      },
    },

  }
)


export default Router
```

**Comment by tsjason** [comment]
@RichardLindhout  I believe you meant `gesturesEnabled: false` :)

**Comment by RichardLindhout** [comment]
@tsjason Good catch, I disabled it later that day ;)

**Comment by aqnaruto** [comment]
does it is work on react-native 0.44,android? why i test it ,it is not work? @RichardLindhout

**Pull request merged**